gen_dict.py

```diff
@@ -20,14 +22,65 @@ def hash_djb2(string):
     return hashval.value & 0x7FFFFFFF
 
 
+def gen_str_dict(code_dir, ignore_some=False, ignore_localized=False):
+    fileglob_list = []
+    str_dict = {}
+
+    pbllog_regex = "\"(?P<loc>[^\)\n]+)\""
+
+    if ignore_some:
+        pblapplog_regex = "(APP_LOG|#define|#import|#include|\/\/|snprintf).*\"(?P<loc>[^\)\n]+)\""
+        pbllog_regex = ".*\"(?P<loc>[^\)\n]+)\""
+    if ignore_localized:
+        pbllocalized_regex = "_\(\"(?P<loc>[^\)\n]+)\"\)"
+
+    for root, dirnames, filenames in os.walk(code_dir):
+        for filename in [filename for filename in filenames if ".c" in filename[-2:]]:
+            fileglob_list.append(os.path.join(root, filename))
+    for filename in fileglob_list:
+        if ignore_some:
+            applog_locs = []
+        if ignore_localized:
+            localized_locs = []
+        str_dict[filename] = []
+        with open(filename, 'rb') as afile:
+            text = afile.read()
+            if ignore_some:
+                applog_match_list = re.finditer(pblapplog_regex, text)
+                if applog_match_list:
+                    for match in applog_match_list:
+                        applog_locs += [match.group('loc')]
+
+            if ignore_localized:
+                localized_match_list = re.finditer(pbllocalized_regex, text)
+                if localized_match_list:
+                    for match in localized_match_list:
+                        localized_locs += [match.group('loc')]
+
+            match_list = re.finditer(pbllog_regex, text)
+            if match_list:
```

Comment:
I'm aware this code is quite complicated, and perhaps unnecessarily so. However, this was the simplest way to only remove single duplicates from the list of unlocalized strings (otherwise, `APP_LOG(APP_LOG_LEVEL_DEBUG, "Hi"); const char *text =  "Hi"; const char *text2 =  _("Hi");` would return an empty list though it should return ["Hi"] from the first declaration in that snippet.)

I also couldn't find a much better way to solve newlines... I'm simply checking for the existence of `APP_LOG` / `#define` / `#include`, etc instead of making sure that this is in the actual function. However, this probably doesn't really matter anyway since the output of this function is only used to inform the user of things they might want to check out.